### PR TITLE
8207025: JvmtiEnv::SetSystemProperty() does not handle OOM

### PR DESCRIPTION
**Pull request by iklam**
`JvmtiEnv::SetSystemProperty` eventually calls `PathString::set_value` in arguments.cpp, which aborts the VM when it fails to allocate a string copy of the property value.

```
bool PathString::set_value(const char *value) {
  if (_value != NULL) {
    FreeHeap(_value);
  }
  _value = AllocateHeap(strlen(value)+1, mtArguments   );
  // should pass AllocFailStrategy::RETURN_NULL -----^
  assert(_value != NULL, "Unable to allocate space for new path value");
```

This should be fixed so that `JvmtiEnv::SetSystemProperty` can return `JVMTI_ERROR_OUT_OF_MEMORY` in case of OOM. See https://docs.oracle.com/en/java/javase/17/docs/specs/jvmti.html#SetSystemProperty

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8207025](https://bugs.openjdk.java.net/browse/JDK-8207025): JvmtiEnv::SetSystemProperty() does not handle OOM

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to 14a44f6388a99f00f5e1da2abebb18a190e45522
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to 14a44f6388a99f00f5e1da2abebb18a190e45522

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7981/head:pull/7981` \
`$ git checkout pull/7981`

Update a local copy of the PR: \
`$ git checkout pull/7981` \
`$ git pull https://git.openjdk.java.net/jdk pull/7981/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7981`

View PR using the GUI difftool: \
`$ git pr show -t 7981`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7981.diff">https://git.openjdk.java.net/jdk/pull/7981.diff</a>

</details>

### CODE REVIEW

**Comment by dholmes-ora** [comment]
Hi Ioi,

I think the real bug in this code is that

`_value = AllocateHeap(strlen(value)+1, mtArguments);`

should be:

`_value = AllocateHeap(strlen(value)+1, mtArguments, AllocFailStrategy::RETURN_NULL);`

as this code is used from `JvmtiEnv::SetSystemProperty` and it seems wrong to abort the VM on that path.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
See comment

**Comment by iklam** [comment]
@dholmes-ora To get `JvmtiEnv::SetSystemProperty` working properly, the fix is more complicated. According to the specification

https://docs.oracle.com/en/java/javase/17/docs/specs/jvmti.html#SetSystemProperty

 `JvmtiEnv::SetSystemProperty`  should return `JVMTI_ERROR_NOT_AVAILABLE` when the property is not available or is not writeable, and JVMTI_ERROR_OUT_OF_MEMORY when OOM (this is one of the "universal errors").

Therefore, I had to change `SystemProperty::set_writeable_value`  to return a tri-state result.

I also added the `JVMTI_ERROR_NULL_POINTER` that was in the spec but wasn't implemented.

@sspitsyn could you take a look as well?

File: src/hotspot/share/prims/jvmtiEnv.cpp [hunk 1]
```diff
@@ -3427,14 +3428,23 @@ JvmtiEnv::GetSystemProperty(const char* property, char** value_ptr) {
 // value - NULL is a valid value, must be checked
 jvmtiError
 JvmtiEnv::SetSystemProperty(const char* property, const char* value_ptr) {
-  jvmtiError err =JVMTI_ERROR_NOT_AVAILABLE;
+  NULL_CHECK(property, JVMTI_ERROR_NULL_POINTER);
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/jvmtiEnv.cpp:3431]: You don't need this. The null check happens in the wrapper code generated from the xml file. See  build/<config>/hotspot/variant-server/gensrc/jvmtifiles/jvmtiEnterTrace.cpp

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Hi Ioi,

I don't think this needs to be that complicated. In set_system_property you can query if the flag is writeable() directly, so then you don't need a ternary return value from set_property_value().

Also see comment below.

David

File: src/hotspot/share/prims/jvmtiEnv.cpp (hunk 1 above)
Comment (sspitsyn) [reply src/hotspot/share/prims/jvmtiEnv.cpp:3431]: I agree with David here that this change is not needed.
This is a relevant fragment from the jvmtiEnter.cpp:
```
static jvmtiError JNICALL
jvmti_SetSystemProperty(jvmtiEnv* env,
            const char* property,
            const char* value_ptr) {
  . . .
  if (Threads::number_of_threads() != 0) {
    . . .
    if (property == NULL) {
      return JVMTI_ERROR_NULL_POINTER;
    }
    err = jvmti_env->SetSystemProperty(property, value_ptr);
  } else {
    if (property == NULL) {
      return JVMTI_ERROR_NULL_POINTER;
    }
    err = jvmti_env->SetSystemProperty(property, value_ptr);
  }
  return err;

```

File: src/hotspot/share/runtime/arguments.hpp [hunk 2]
```diff
@@ -107,8 +108,9 @@ class SystemProperty : public PathString {
   bool internal() const               { return _internal; }
   SystemProperty* next() const        { return _next; }
   void set_next(SystemProperty* next) { _next = next; }
+  bool writeable() const              { return _writeable; }
 
-  bool is_readable() const {
+  bool readable() const {
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/arguments.hpp:113]: Might be better/simpler to keep is_readable and change to is_writable(), as that avoids changing other files.

Comment (iklam) [reply src/hotspot/share/runtime/arguments.hpp:113]: There's also `internal()`, which would need to be changed to `is_internal()` as well for consistency. I chose to change `is_readable()` to `readable()` so I just needed to change one name.

Comment (dholmes-ora) [reply src/hotspot/share/runtime/arguments.hpp:113]: Okay.

**Review by sspitsyn — APPROVED** [review]
Hi Ioi,
The fix looks good to me. I leave the decision about is_readable/readable to you and David. I have a very little preference to is_* variant for consistency with other places in HopSpot I'm aware off.
Thank you a lot for taking care about this!
-Serguei

**Review by dholmes-ora — APPROVED** [review]
On the JVMTI side this looks good and corrects an oversight.

The SystemProperty API changes are not quite what I expected but I see why you did them.

Thanks,
David

**Comment by iklam** [comment]
Thanks @dholmes-ora and @sspitsyn for the review.
/integrate

**Review by alexmenkov — APPROVED** [review]

**Pull request closed without merging**